src/main/java/io/cdap/plugin/gcp/gcs/connector/GCSConnector.java

```diff
@@ -0,0 +1,234 @@
+      storage.list(Storage.BucketListOption.pageSize(1));
+    } catch (Exception e) {
+      failureCollector.addFailure(String.format("Not able to connect to GCS. Error: %s", e.getMessage()),
+                                  "Please provide valid configs to connect to GCS");
+    }
+  }
+
+  @Override
+  public BrowseDetail browse(ConnectorContext connectorContext, BrowseRequest request) throws IOException {
+    String path = request.getPath();
+    int limit = request.getLimit() == null || request.getLimit() <= 0 ? Integer.MAX_VALUE : request.getLimit();
+    if (isRoot(path)) {
+      return browseBuckets(limit);
+    }
+    return browseBlobs(GCSPath.from(path), limit);
+  }
+
+  @Override
+  protected String getFullPath(String path) {
+    if (isRoot(path)) {
+      return GCSPath.SCHEME;
+    }
+    return GCSPath.from(path).getUri().toString();
+  }
+
+  @Override
+  protected Map<String, String> getFileSystemProperties(String path) {
+    return GCPUtils.getFileSystemProperties(config, path, new HashMap<>());
+  }
+
+  @Override
+  protected void setConnectorSpec(ConnectorSpecRequest request, ConnectorSpec.Builder builder) {
+    super.setConnectorSpec(request, builder);
+    builder.addRelatedPlugin(
+      new PluginSpec(GCSSource.NAME, BatchSource.PLUGIN_TYPE,
+                     ImmutableMap.of(
+                       GCSSource.GCSSourceConfig.NAME_USE_CONNECTION, "true",
+                       GCSSource.GCSSourceConfig.NAME_CONNECTION, request.getConnectionWithMacro(),
+                       GCSSource.GCSSourceConfig.NAME_PATH, getFullPath(request.getPath()))));
+  }
+
+  private BrowseDetail browseBuckets(int limit) throws IOException {
+    Storage storage = getStorage();
+    Page<Bucket> buckets = storage.list();
+    int count = 0;
+    BrowseDetail.Builder builder = BrowseDetail.builder();
+    for (Bucket bucket : buckets.iterateAll()) {
+      if (count < limit) {
+        String name = bucket.getName();
+        builder.addEntity(BrowseEntity.builder(name, name, BUCKET_TYPE).canBrowse(true).canSample(true).build());
+      }
+      count++;
+    }
+    return builder.setTotalCount(count).build();
+  }
+
+  private BrowseDetail browseBlobs(GCSPath path, int limit) throws IOException {
+    Storage storage = getStorage();
+    String pathBlobName = path.getName();
+    Page<Blob> blobs = storage.list(path.getBucket(), Storage.BlobListOption.currentDirectory(),
+                                    Storage.BlobListOption.prefix(pathBlobName));
+    int count = 0;
+    BrowseDetail.Builder builder = BrowseDetail.builder();
+    // entity for the path itself will also get returned in the result since this is a prefix search.
+    BrowseEntity entityForPath = null;
+    for (Blob blob : blobs.iterateAll()) {
+      String blobName = blob.getName();
+
+      if (count >= limit) {
+        if (!blobName.equals(pathBlobName)) {
+          count++;
+        }
+        continue;
```

Comment:
why not break here?

Reply:
Because we still need to count to get the totalCount

Reply:
Discussed offline, we will not continue counting for the number to avoid the request hang for large number of files, so changed to break here